wrappers/services.go

```diff
@@ -75,30 +76,109 @@ func generateSnapServiceFile(app *snap.AppInfo, opts *AddSnapServicesOptions) ([
 	return genServiceFile(app, opts)
 }
 
+// max returns the maximum of two integers. Why is this
+// not provided by golang?
+func max(a, b int) int {
+	if a < b {
+		return b
+	}
+	return a
+}
+
+func min(a, b int) int {
+	if b < a {
+		return b
+	}
+	return a
+}
+
+func formatCpuGroupSlice(grp *quota.Group) string {
+	if grp.CPULimit == nil {
+		return ""
+	}
+
+	// calculateSystemdCPULimit calculates the number of cpus and the allowed percentage
+	// to the systemd specific format. The CPUQuota setting is only available since systemd 213
+	calculateSystemdCPULimit := func() string {
+		if grp.CPULimit.Count == 0 && grp.CPULimit.Percentage == 0 {
+			return ""
+		}
+
+		cpuQuotaSnap := max(grp.CPULimit.Count, 1) * grp.CPULimit.Percentage
+		cpuQuotaMax := runtime.NumCPU() * 100
+		return fmt.Sprintf("CPUQuota=%d%%\n", min(cpuQuotaSnap, cpuQuotaMax))
+	}
+
+	// getAllowedCpusValue converts allowed CPUs array to a comma-delimited
+	// string that systemd expects it to be in. If the array is empty then
+	// an empty string is returned
+	getAllowedCpusValue := func() string {
+		if len(grp.CPULimit.AllowedCPUs) == 0 {
+			return ""
+		}
+
+		allowedCpusValue := strutil.SliceToCommaSeparatedString(grp.CPULimit.AllowedCPUs)
+		return fmt.Sprintf("AllowedCPUs=%s\n", allowedCpusValue)
+	}
+
+	template := `# Always enable cpu accounting, so the following cpu quota options have an effect
+CPUAccounting=true
+`
+	return fmt.Sprint(template, calculateSystemdCPULimit(), getAllowedCpusValue(), "\n")
+}
+
+func formatMemoryGroupSlice(grp *quota.Group) string {
+	buf := bytes.Buffer{}
+	template := `# Always enable memory accounting otherwise the MemoryMax setting does nothing.
+MemoryAccounting=true
+`
+	fmt.Fprint(&buf, template)
+
+	if grp.MemoryLimit != 0 {
+		valuesTemplate := `MemoryMax=%[1]d
+# for compatibility with older versions of systemd
+MemoryLimit=%[1]d
+
+`
+		memoryValues := fmt.Sprintf(valuesTemplate, grp.MemoryLimit)
+		fmt.Fprint(&buf, memoryValues)
+	}
+	return buf.String()
+}
+
+func formatTaskGroupSlice(grp *quota.Group) string {
+	buf := bytes.Buffer{}
+	template := `# Always enable task accounting in order to be able to count the processes/
```

Comment:
this is more of a header than a template

Reply:
agreed, i'll get them changed